### PR TITLE
python312Packages.pytorch-lightning: 2.3.3 -> 2.4.0

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Changelog: https://github.com/Lightning-AI/pytorch-lightning/releases/tag/refs/tags/2.4.0

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
We need to wait for the next release of [finetuning-scheduler](https://github.com/speediedan/finetuning-scheduler):
```
Checking runtime dependencies for finetuning_scheduler-2.3.3-py3-none-any.whl
  - pytorch-lightning<2.3.4,>=2.3.0 not satisfied by version 2.4.0
```

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
> Result of `nixpkgs-review pr 333415` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

Fails because `faiss` is broken.
This is already the case on `master` and thus not a regression.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 333415` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>python311Packages.finetuning-scheduler</li>
    <li>python311Packages.finetuning-scheduler.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pytorch-lightning</li>
    <li>python311Packages.pytorch-lightning.dist</li>
    <li>python312Packages.finetuning-scheduler</li>
    <li>python312Packages.finetuning-scheduler.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-lightning</li>
    <li>python312Packages.pytorch-lightning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
